include/nbl/builtin/hlsl/sort/counting.hlsl

```diff
@@ -0,0 +1,137 @@
+// Copyright (C) 2018-2024 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#include "nbl/builtin/hlsl/sort/common.hlsl"
+#include "nbl/builtin/hlsl/workgroup/arithmetic.hlsl"
+
+#ifndef _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+#define _NBL_BUILTIN_HLSL_SORT_COUNTING_INCLUDED_
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace sort
+{
+
+template<
+    uint16_t GroupSize,
+    uint16_t KeyBucketCount,
+    typename Key,
+    typename KeyAccessor,
+    typename ValueAccessor,
+    typename ScratchAccessor,
+    typename SharedAccessor,
+    bool robust=false
+>
+struct counting
+{
+    uint32_t inclusive_scan(uint32_t value, NBL_REF_ARG(SharedAccessor) sdata)
+    {
+        return workgroup::inclusive_scan < plus < uint32_t >, GroupSize >::
+                template __call <SharedAccessor>(value, sdata);
+    }
+
+    void build_histogram(NBL_REF_ARG( KeyAccessor) key, NBL_REF_ARG(SharedAccessor) sdata, const CountingParameters<Key> data)
+    {
+        uint32_t tid = workgroup::SubgroupContiguousIndex();
+        uint32_t buckets_per_thread = (KeyBucketCount + GroupSize - 1) / GroupSize;
+        uint32_t baseIndex = (glsl::gl_WorkGroupID().x * GroupSize) * data.elementsPerWT;
```

Comment:
relying on the HW/API dispatch being 1D is kinda flaky, because:
1. One could do the histogramming pass from a 2D Dispatch (e.g. Visibility Buffer Material ID Binning)
2. There is a fundamental limit to X-axis of the dispatch which you'd run into when trying to sort more than 64M keys on some GPUs -> http://vulkan.gpuinfo.org/displaydevicelimit.php?name=maxComputeWorkGroupCount[0]&platform=all

The solution is to allow for "virtual"/"persistent" workgroups, and the used to feed you a `workGroupIndex` instead of relying on `gl_WorkGroupID`

Reply:
Done